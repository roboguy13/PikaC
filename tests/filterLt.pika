%generate filter

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

filter : Int -> Sll -> Sll;
filter n Nil := Nil [Sll];
filter n (Cons a as) | a < n := filter n as;
filter n (Cons a as) | not (a < n) := Cons a (filter n as) [Sll];

%test "filter1" Sll:
  filter 2 (Cons 1 (Cons 2 (Cons 10 (Cons 20 (Nil [Sll]) [Sll]) [Sll]) [Sll]) [Sll]);

