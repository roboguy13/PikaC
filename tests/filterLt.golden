void filter(loc i1, loc x, loc x2);
void filter(loc i1, loc x, loc x2) {
 if (x == 0) {
  WRITE_LOC(x2, 0, 0);
 } else {
  loc h = READ_LOC(x, 0);
  loc n = READ_LOC(x, 1);
  if (i1 <= h) {
   filter(i1, n, x2);
   loc x4 = READ_LOC(x2, 0);
   loc x = (loc)malloc(2 * sizeof(loc));
   WRITE_LOC(x2, 0, x);
   WRITE_LOC(x, 1, x4);
   WRITE_LOC(x, 0, h);
  } else {
   filter(i1, n, x2);
  }
 }
}